### ISSUE TITLE
about normalize_adj

### ISSUE THREAD

**Issue by srtianxia** [dmlc/dgl#1285]
## ❓ Questions and Help

In the original author's implementation of gcn, there is a function `normalize_adj`. I want to assign the value of the edge obtained after `normalize_adj` to `edata`(g.edata['norm_edge']= ...). What should I do, thank you!

```python
def normalize_adj(adj):
    """Symmetrically normalize adjacency matrix."""
    adj = sp.coo_matrix(adj)
    rowsum = np.array(adj.sum(1))
    d_inv_sqrt = np.power(rowsum, -0.5).flatten()
    d_inv_sqrt[np.isinf(d_inv_sqrt)] = 0.
    d_mat_inv_sqrt = sp.diags(d_inv_sqrt)
    return adj.dot(d_mat_inv_sqrt).transpose().dot(d_mat_inv_sqrt).tocoo()
```

**Comment by mufeili** [comment]
See the code snippet below.

```python
import dgl
import dgl.function as fn
import numpy as np
import scipy.sparse as sp
import torch

edge_list = [(0, 1), (1, 0), (1, 2), (2, 1), (1, 3), (3, 1), (0, 3), (3, 0)]
g = dgl.DGLGraph(edge_list)
degs = g.in_degrees().float()
norm = torch.pow(degs, -0.5)
norm[torch.isinf(norm)] = 0
g.ndata['norm'] = norm.unsqueeze(1)
g.apply_edges(fn.u_mul_v('norm', 'norm', 'normalized'))
print(g.edata['normalized'])

def normalize_adj(adj):
    """Symmetrically normalize adjacency matrix."""
    adj = sp.coo_matrix(adj)
    rowsum = np.array(adj.sum(1))
    d_inv_sqrt = np.power(rowsum, -0.5).flatten()
    d_inv_sqrt[np.isinf(d_inv_sqrt)] = 0.
    d_mat_inv_sqrt = sp.diags(d_inv_sqrt)
    return adj.dot(d_mat_inv_sqrt).transpose().dot(d_mat_inv_sqrt).tocoo()

normalized_adj = normalize_adj(g.adjacency_matrix().to_dense()).toarray()
normalized_values = [normalized_adj[e] for e in edge_list]
print(normalized_values)
```

**Comment by srtianxia** [comment]
Wow~ cool. thanks a lot !

**Comment by srtianxia** [comment]
@mufeili Hi, Thank you for your help yesterday, But what to do with this `si_normalize_adj`, thank you!
```python
def si_normalize_adj(adj):
    adj = sp.coo_matrix(adj)

    rowsum = np.array(adj.sum(1))

    d_inv = np.power(rowsum, -1).flatten()
    d_inv[np.isinf(d_inv)] = 0.
    d_mat_inv = sp.diags(d_inv)

    norm_adj = d_mat_inv.dot(adj)
    return norm_adj.tocoo()
```

**Comment by mufeili** [comment]
> @mufeili Hi, Thank you for your help yesterday, But what to do with this `si_normalize_adj`, thank you!
> 
> ```python
> def si_normalize_adj(adj):
>     adj = sp.coo_matrix(adj)
> 
>     rowsum = np.array(adj.sum(1))
> 
>     d_inv = np.power(rowsum, -1).flatten()
>     d_inv[np.isinf(d_inv)] = 0.
>     d_mat_inv = sp.diags(d_inv)
> 
>     norm_adj = d_mat_inv.dot(adj)
>     return norm_adj.tocoo()
> ```
> 
> My problem is mainly in `g.apply_edges (fn.u_mul_v ('norm', 'norm', 'normalized'))` I do not know how to operate in `si_normalize_adj`

For this case, I will do something as follows:
```python
import dgl
import torch

def copy_v(edges):
    return {'he': edges.dst['norm']}

edge_list = [(0, 1), (1, 0), (1, 2), (2, 1), (1, 3), (3, 1), (0, 3), (3, 0)]
g = dgl.DGLGraph(edge_list)
degs = g.in_degrees().float()
norm = torch.pow(degs, -1)
norm[torch.isinf(norm)] = 0
g.ndata['norm'] = norm.unsqueeze(1)
g.apply_edges(copy_v)
print(g.edata['he'])
```

**Comment by mufeili** [comment]
> @mufeili Hi, Thank you for your help yesterday, But what to do with this `si_normalize_adj`, thank you!
> 
> ```python
> def si_normalize_adj(adj):
>     adj = sp.coo_matrix(adj)
> 
>     rowsum = np.array(adj.sum(1))
> 
>     d_inv = np.power(rowsum, -1).flatten()
>     d_inv[np.isinf(d_inv)] = 0.
>     d_mat_inv = sp.diags(d_inv)
> 
>     norm_adj = d_mat_inv.dot(adj)
>     return norm_adj.tocoo()
> ```
> 
> My problem is mainly in `g.apply_edges (fn.u_mul_v ('norm', 'norm', 'normalized'))` I do not know how to operate in `si_normalize_adj`

For this case, I will do something as follows:
```python
import dgl
import torch

def copy_v(edges):
    return {'he': edges.dst['norm']}

edge_list = [(0, 1), (1, 0), (1, 2), (2, 1), (1, 3), (3, 1), (0, 3), (3, 0)]
g = dgl.DGLGraph(edge_list)
degs = g.in_degrees().float()
norm = torch.pow(degs, -1)
norm[torch.isinf(norm)] = 0
g.ndata['norm'] = norm.unsqueeze(1)
g.apply_edges(copy_v)
print(g.edata['he'])
```

**Comment by srtianxia** [comment]
@mufeili  Thanks a lot!
